File: Easy/validate-subsequence.py

"""  
Given two non-empty arrays of integers, write a function that determines whether the second array is a subsequence of the first one.
"""

def isValidSubsequence(array, sequence):
    # Write your code here.
    # O(N) time and O(1) space
    if not sequence:
		return False
	
	index = 0
	for x in array:
		if x == sequence[index]:
			if index == len(sequence) - 1:
				return True
			index += 1
			
	return False

File: Easy/kotlin/validate-subsequence.kt
@file:JvmName("Main")
fun main()
{
    println("result is ${isValidSubsequence(intArrayOf(5,1,2,3,4),intArrayOf(1,2,3))}")
}

// O(N) time and O(1) space
fun isValidSubsequence(array: IntArray, sequence: IntArray) : Boolean
{
        var i = 0
        var sequenceStarted = false
        for(item in array)
        {
            if(item == sequence[i])
            {
               sequenceStarted = true
               i++

               if(i == sequence.size) return true

            } else if (sequenceStarted) return false
        }

    return sequenceStarted
}